compiler/rustc_middle/src/mir/interpret/allocation/init_mask.rs

```diff
@@ -10,20 +10,185 @@ type Block = u64;
 
 /// A bitmask where each bit refers to the byte with the same index. If the bit is `true`, the byte
 /// is initialized. If it is `false` the byte is uninitialized.
-// Note: for performance reasons when interning, some of the `InitMask` fields can be partially
-// hashed. (see the `Hash` impl below for more details), so the impl is not derived.
-#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable)]
-#[derive(HashStable)]
+/// The actual bits are only materialized when needed, and we try to keep this data lazy as long as
+/// possible. Currently, if all the blocks have the same value, then the mask represents either a
+/// fully initialized or fully uninitialized const allocation, so we can only store that single
+/// value.
+#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, Hash, HashStable)]
 pub struct InitMask {
-    blocks: Vec<Block>,
+    pub(crate) blocks: InitMaskBlocks,
     len: Size,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, Hash, HashStable)]
+pub(crate) enum InitMaskBlocks {
```

Comment:
Why do so many things here have to be public to the entire crate?

Reply:
For the unit tests in the `tests` module. They look at the mask structure to count the number of actually materialized blocks. It seemed cleaner and harder to break accidentally, than to have a `#[cfg(test)]` method returning the count ? But if you'd prefer that, either is fine by me.

Reply:
You can put the tests into `interpret/allocation/init_mask/tests.rs`, then they can access private fields without having to make them public.

Reply:
I'm aware, but there were existing init mask tests in the other module and didn't really want to change existing things too much here. In any case if you prefer that I've pushed a commit to do so.

I'll let you notify bors if you're happy.